### PR TITLE
MatMul Gradient optimization for dB when B's is 2D tensor

### PR DESCRIPTION
**Pull request by SherlockNoMad**
**Description**: 
Y = MatMul(A, B)
In general form: dB = ReduceSum(A' * dY) 
For example A[500, 4, 1024], B[1024, 1024], Y[500, 4, 1024]. 
Pre-ReduceSum Tensor = A' * dY => [500, 1024, 1024], which could be very large (2GB in fp32). 
However, this temporary tensor is immediately latter reduced-sum into [1024, 1024]. 
This fix is meant to avoid generating the large intermediate tensor. 

This is achieved by reshaping A and dY into 2D tensors first, A => [500x4, 1024], dY => [500x4, 1024] and dB can be obtained without ReduceSum and intermediate tensor. 

However, similar trick **cannot** be applied for dA calculation. 
For example, A[1024, 1024], B[500, 1024, 4], Y[500, 1024, 4]. 
dA = ReduceSum(dY * B')
The collapsing dimensions (500 and 4) for dY and B are not adjacent to each other, so a simple Reshape is not working here. 
If we encounter such case in the future, we will need another fix.

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -317,17 +321,54 @@ IMPLEMENT_GRADIENT_BUILDER(GetMatMulGradient) {
...
+            one_const_node,
+            neg_one_const_node,
+            zero_float_const_node,
+
+            NodeDef("Shape", {B}, {IA("B_shape")}),
+
+            // reshape A to 2D [M, K]
+            NodeDef("Gather", {IA("B_shape"), ZERO_I}, {IA("K_dim")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Concat", {NEG_ONE, IA("K_dim")}, {IA("A_target_shape")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Reshape", {A, IA("A_target_shape")}, {IA("A_reshape_2d")}),
```
Comment (ashbhandare) [inline orttraining/orttraining/core/graph/gradient_builder.cc:347]: Can you just reshape it to [-1 0] ?

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 2]
```diff
@@ -317,17 +321,54 @@ IMPLEMENT_GRADIENT_BUILDER(GetMatMulGradient) {
...
+
+            // reshape dY to 2D [M, N]
+            NodeDef("Gather", {IA("B_shape"), ONE}, {IA("N_dim")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Concat", {NEG_ONE, IA("N_dim")}, {IA("dY_target_shape")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Reshape", {GO(0), IA("dY_target_shape")}, {IA("dY_reshape_2d")}),
```
Comment (ashbhandare) [inline orttraining/orttraining/core/graph/gradient_builder.cc:352]: same here, reshape to [-1,0]?

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:347]: no... reshaping [500, 4, 1024] with [-1, 0] will end in [500x1024, 4]. 0 is indicating following left aligned shape. 
We need target shape [500x4, 1024].

Comment (KeDengMS) [reply orttraining/orttraining/core/graph/gradient_builder.cc:347]: Squeeze(Reshape(x, [1,-1,0]), 0) should do.

---
In reply to: [475955739](https://github.com/microsoft/onnxruntime/pull/4899#discussion_r475955739) [](ancestors = 475955739)

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:347]: x could be arbitrary dimension. eg. for 4D x, you would need [1,1,-1,0] instead.

Comment (KeDengMS) [reply orttraining/orttraining/core/graph/gradient_builder.cc:347]: Flatten(x, axis=-1)?
[https://github.com/onnx/onnx/blob/master/docs/Operators.md#Flatten](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Flatten)

---
In reply to: [475967015](https://github.com/microsoft/onnxruntime/pull/4899#discussion_r475967015) [](ancestors = 475967015)

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc:347]: Thanks a lot!!! Flatten will do the trick here!

**Comment by SherlockNoMad** [comment]
It turns out that Flatten op that supports negative axis comes in opset 11. If I use this op for MatMulGrad, we are losing support for model with opset < 11. Unfortunately, we still have an ELR model is using opset 10. 

So I have to save this PR for the future. I am reverting back to a earlier commit where I didn't use Flatten op....

**Pull request merged**